tensorflow_addons/optimizers/adabelief.py

```diff
@@ -0,0 +1,353 @@
+        self.rectify = rectify
+        self._has_weight_decay = weight_decay != 0.0
+        self._initial_total_steps = total_steps
+
+    def _create_slots(self, var_list):
+        for var in var_list:
+            self.add_slot(var, "m")
+        for var in var_list:
+            self.add_slot(var, "v")
+        if self.amsgrad:
+            for var in var_list:
+                self.add_slot(var, "vhat")
+
+    def set_weights(self, weights):
+        params = self.weights
+        num_vars = int((len(params) - 1) / 2)
+        if len(weights) == 3 * num_vars + 1:
+            weights = weights[: len(params)]
+        super().set_weights(weights)
+
+    def _decayed_wd(self, var_dtype):
+        wd_t = self._get_hyper("weight_decay", var_dtype)
+        if isinstance(wd_t, tf.keras.optimizers.schedules.LearningRateSchedule):
+            wd_t = tf.cast(wd_t(self.iterations), var_dtype)
+        return wd_t
+
+    def _resource_apply_dense(self, grad, var):
+        var_dtype = var.dtype.base_dtype
+        lr_t = self._decayed_lr(var_dtype)
+        wd_t = self._decayed_wd(var_dtype)
+        m = self.get_slot(var, "m")
+        v = self.get_slot(var, "v")
+        beta_1_t = self._get_hyper("beta_1", var_dtype)
+        beta_2_t = self._get_hyper("beta_2", var_dtype)
+        epsilon_t = tf.convert_to_tensor(self.epsilon, var_dtype)
+        local_step = tf.cast(self.iterations + 1, var_dtype)
+        beta_1_power = tf.pow(beta_1_t, local_step)
+        beta_2_power = tf.pow(beta_2_t, local_step)
+
+        if self._initial_total_steps > 0:
+            total_steps = self._get_hyper("total_steps", var_dtype)
+            warmup_steps = total_steps * self._get_hyper("warmup_proportion", var_dtype)
+            min_lr = self._get_hyper("min_lr", var_dtype)
+            decay_steps = tf.maximum(total_steps - warmup_steps, 1)
+            decay_rate = (min_lr - lr_t) / decay_steps
+            lr_t = tf.where(
+                local_step <= warmup_steps,
+                lr_t * (local_step / warmup_steps),
+                lr_t + decay_rate * tf.minimum(local_step - warmup_steps, decay_steps),
+            )
+
+        sma_inf = 2.0 / (1.0 - beta_2_t) - 1.0
+        sma_t = sma_inf - 2.0 * local_step * beta_2_power / (1.0 - beta_2_power)
+
+        m_t = m.assign(
+            beta_1_t * m + (1.0 - beta_1_t) * grad, use_locking=self._use_locking
+        )
+        m_corr_t = m_t / (1.0 - beta_1_power)
+
+        v_t = v.assign(
+            beta_2_t * v + (1.0 - beta_2_t) * tf.math.square(grad - m_t) + epsilon_t,
+            use_locking=self._use_locking,
+        )
+        if self.amsgrad:
+            vhat = self.get_slot(var, "vhat")
+            vhat_t = vhat.assign(tf.maximum(vhat, v_t), use_locking=self._use_locking)
+            v_corr_t = tf.sqrt(vhat_t / (1.0 - beta_2_power))
+        else:
+            vhat_t = None
+            v_corr_t = tf.sqrt(v_t / (1.0 - beta_2_power))
+
+        r_t = tf.sqrt(
```

Comment:
Here the readability is completely gone. Can you try to make it more readable, split the conditions?

Reply:
Yeah, thanks for the review. The ```r_t``` computation is actually forked from ```rectified_adam.py```, and I just moved it within the ```if self.rectify``` condition, the idea is to use recitified learning rate as in RAdam when ```self.rectify=True```

For the computation of ```r_t```, do you prefer to keep the current format so that is matches ```RAdam``` exactly, or I break it into shorter lines?

Reply:
I prefer readability. Let's try to break it in a two lines or so

Reply:
Break it into 3 lines to match the numerator, denominator and sqrt operation as in line 11 of Algo2 in https://arxiv.org/pdf/1908.03265.pdf
Please take a look if that reads better now.

Reply:
PS, the long-lines of ```r_t```  is also in ```rectified_adam``` in tensorflow_addons, https://github.com/tensorflow/addons/blob/337951b2371dfa2bc3d4d5cd778de9f25b6f0789/tensorflow_addons/optimizers/rectified_adam.py#L209
Perhaps we can consider format that part too for readability.

Reply:
Yes we should